### ISSUE TITLE
Thoughts about converting to a ansible-galaxy collection?

### ISSUE THREAD

**Issue by propyless** [willshersystems/ansible-sshd#176]
Hello!

So I was wondering if theres any interest to convert the role to a role in a collection instead?

**Comment by Jakuje** [comment]
We are shipping this as part of "linux system roles" collection in the galaxy:

https://galaxy.ansible.com/fedora/linux_system_roles

Its a question if it is worth converting this role into a separate collection, but it should not be that hard if it would be needed.

**Comment by mattwillsher** [comment]
What are the benefits of a seperate collection?

**Comment by rehanone** [comment]
> What are the benefits of a seperate collection?

The name itself is reason enough. the collection is currently called `fedora.linux_system_roles`. Since I have no use for fedora as a server. that is the last place anyone would go looking for an `sshd` role to use on Debian/Ubuntu. It took me a day to finally find this issue and get the collection downloaded. I would agree with @propyless to move this role into a separate collection and maybe update the documentation on how to use it.

**Comment by richm** [comment]
> > What are the benefits of a seperate collection?
> 
> The name itself is reason enough. the collection is currently called `fedora.linux_system_roles`. Since I have no use for fedora as a server.

I agree that the name is misleading.  The role/collection works on a wide variety of platforms, not only Fedora.

> that is the last place anyone would go looking for an `sshd` role to use on Debian/Ubuntu. It took me a day to finally find this issue and get the collection downloaded. I would agree with @propyless to move this role into a separate collection and maybe update the documentation on how to use it.

Yes, there could be a collection `willshershsytems.ansible_sshd` which contains only the role `sshd`.  We will still keep the role in `fedora.linux_system_roles` because we need it there, but there could be multiple collections which provide this role.

**Comment by jose-d** [comment]
wouldn't be good to mention `fedora.linux_system_roles` collection somehow at the main Readme.md ?
As I read the note from @rehanone above, I just quickly refactored my tooling to include whole `fedora.linux_system_roles` - I had no clue that this role is part of this collection.

**Comment by richm** [comment]
> wouldn't be good to mention `fedora.linux_system_roles` collection somehow at the main Readme.md ?

Maybe?  The role is provided as a legacy role in the legacy role format - how is it supposed to know if other collections include it?

I get it - it is frustrating when you want to/must use collections and cannot use the legacy role format, but that isn't the "fault" of the ansible-sshd role being provided in the legacy role format.

If there is a huge demand to provide a collection instead of the legacy role format, the maintainers of the role (or anyone else, really) can provide such a collection.

> As I read the note from @rehanone above, I just quickly refactored my tooling to include whole `fedora.linux_system_roles` - I had no clue that this role is part of this collection.

Or you can use the legacy role format (unless you must use a collection, or just really don't want to use the legacy role format).

**Comment by jose-d** [comment]
> Or you can use the legacy role format (unless you must use a collection, or just really don't want to use the legacy role format).

I suspect people (including myself) are pushed in direction of using collections instead simple roles because _Ansible executive environments_, used by _AWX_ don't support adding individual roles easily..

**Comment by rehanone** [comment]
> 
> Yes, there could be a collection `willshershsytems.ansible_sshd` which contains only the role `sshd`. We will still keep the role in `fedora.linux_system_roles` because we need it there, but there could be multiple collections which provide this role.

I think is the best outcome. It will allow legacy users to continue using `fedora.linux_system_roles` and for new users give a clean collection to work with.

**Comment by domrim** [comment]
this seems to get more and more important. any progress on this? is there assistance needed/wanted for migrating the role into a collection?

**Comment by richm** [comment]
@mattwillsher not sure what you want to do here.  If you want to keep the github repo in the "legacy" role format, but publish the role as a collection (basically, just "wrapped" in a collection), then you could do what we do for the system roles - we have a script to convert the role to collection format, then publish to Galaxy as a collection.  We also have a github action to automate this whenever there is a new release of a role - see https://github.com/linux-system-roles/auto-maintenance/actions/workflows/publish_collection.yml - you could do something similar for ansible-sshd

**Comment by mattwillsher** [comment]
@richm thats all sounds fine to me.

**Comment by mattwillsher** [comment]
@richm can you provide the script thatbdoes tge conversion along with pointrr to good practice (e.g. is advided to go to bump the major version?)
Otherwise, Im happy for soneone in your team to apply the conversion.

**Comment by richm** [comment]
> [@richm](https://github.com/richm) can you provide the script thatbdoes tge conversion along with pointrr to good practice (e.g. is advided to go to bump the major version?) Otherwise, Im happy for soneone in your team to apply the conversion.

The "top level" script is here - https://github.com/linux-system-roles/auto-maintenance/blob/main/release_collection.py
This is the main driver for the publish_collection.yml workflow I posted above.
This script in turn uses https://github.com/linux-system-roles/auto-maintenance/blob/main/lsr_role2collection.py to do the low level role-to-collection conversion.
There are a couple of config files used - https://github.com/linux-system-roles/auto-maintenance/blob/main/collection_release.yml has the information about the latest org/repo/tag used in the collection - https://github.com/linux-system-roles/auto-maintenance/blob/main/galaxy.yml is required by the Ansible collection, as well as the files in https://github.com/linux-system-roles/auto-maintenance/tree/main/lsr_role2collection

System roles follows strict semantic versioning X.Y.Z, and the release_collection.py script implements this, in conjunction with the collection_release.yml file - if there is an API or other breaking change in any of the roles (the role X version has increased), then the collection X digit is increased and the Y and Z are reset to 0.  If any of the roles has a new feature (the role Y version has increased), then the collection Y digit is increased and the Z is reset to 0.  Any other change of the Z digit in any role will bump the Z version in the collection.

For sshd - this will be much simpler - there is only 1 role, and you probably want the collection version to be exactly the same as the role version.
